### PR TITLE
A3: Number Conversion 

### PR DESCRIPTION
**Pull request by DhaaraniMohanasundaram**

### CODE REVIEW

File: Training24/NumConversion.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+// ------------------------------------------------------------------------------------------------
+
+#include <stdio.h>
+#include <stdbool.h>
+#include <string.h>
+#include <ctype.h>
+
+// Function to  convert an integer to binary (8-bit: 256 values)
+void ConvertToBinary (int number) {
+   for (int i = 7; i >= 0; i--)    printf ("%d", (number >> i) & 1);
```
Comment (dineshN52) [inline Training24/NumConversion.c:16]: Create common method for printing by passing values of iterator and array

File: Training24/NumConversion.c [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#include <ctype.h>
+
+// Function to  convert an integer to binary (8-bit: 256 values)
+void ConvertToBinary (int number) {
+   for (int i = 7; i >= 0; i--)    printf ("%d", (number >> i) & 1);   
+   printf ("\n");
+}
+
+// Function to  check if input is a valid integer
+bool isValidInteger (const char* str) {
```
Comment (dineshN52) [inline Training24/NumConversion.c:21]: Method name should always start with capital

Comment (dineshN52) [inline Training24/NumConversion.c]: Method name should always start with capital

**Review by dineshN52 — CHANGES_REQUESTED** [review]

File: Training24/NumConversion.c [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+// Numconversion.c
+// Program to get user input(only integer) and convert them to Hexadecimal and Binary.
+// ------------------------------------------------------------------------------------------------
+
+#include <stdio.h>
+#include <stdbool.h>
+#include <string.h>
+#include <ctype.h>
+
+// Function to  convert an integer to binary (8-bit: 256 values)
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Please comment your code properly as previously demonstrated.

File: Training24/NumConversion.c [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+void ConvertToBinary (int number) {
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: You should not print output here, you should only return the comverted value

Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: You should not print output here, you should only return the converted value

File: Training24/NumConversion.c [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// Function to convert an integer to its 8-bit hexadecimal representation using modulus
+void HexUsingModulus (int number) {
+   unsigned char postiveInt = (unsigned char)number;    // To handle negative values
+   char hex[3];   // Array to store hexadecimal representation (2 digits + null terminator)
+   int index = 0;
+   if (postiveInt == 0) hex[index++] = '0';   
+   else {   
+      while (postiveInt > 0) {
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: positiveInt is an unsigned char. It will never be <0. You should use index for your comparison. Also, why do you need this variable type to be an unsigned char? You could work with int/char and it would yield the same result. Also you are assigning an integer value to a char (which can never be signed anyway). Please refrain from using / converting to data types where it is not necessary.

File: Training24/NumConversion.c [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+         int remainder = postiveInt % 16;   // Get the remainder
+         switch (remainder) {    // Convert remainder to ASCII hexadecimal character
+         case 0: case 1: case 2: case 3: case 4: case 5: case 6: case 7: case 8: case 9:
+            hex[index++] = '0' + remainder;
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: `hex[index++] = '0' + remainder;`
number + 0 = number. This is not required. You can just break.

Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: `hex[index++] = '0' + remainder;`
number + 0 = number. This is not required. You can increment and break.

File: Training24/NumConversion.c [hunk 7]
```diff
@@ -0,0 +1,100 @@
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// Numconversion.c
+// Program to get user input(only integer) and convert them to Hexadecimal and Binary.
+// ------------------------------------------------------------------------------------------------
+
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: avoid leaving empty lines before the directives

File: Training24/NumConversion.c [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+      char temp = hex[j];
+      hex[j] = hex[index - j - 1];
+      hex[index - j - 1] = temp;
+   }
+   hex[index] = '\0';   // Ensure null-termination
+   return;
+}
+
+/// <summary>Function to determine the appropriate bit width based on the integer value.</summary>
+int DetermineBitWidth (int number) {
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: This is nice!

File: Training24/NumConversion.c [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+   }
+   return false;
+}
+
+/// <summary>Function to convert an integer into hexadecimal representation.</summary>
+void ConvertToHex (int number, char hex[], int bitWidth) {
+   int value;
+   int hexDigits = bitWidth / 4;   // Calculate number of hex digits based on bit width
+   // Handle negative numbers by computing two's complement
+   if (number < 0) value = (1 << bitWidth) + number;
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: You can use a [ternary operator](https://stackoverflow.com/questions/392932/how-do-i-use-the-conditional-ternary-operator) here
```suggestion
   value = number < 0 ? (1 << bitWidth) + number : number;
```

File: Training24/NumConversion.c [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+   else value = number;
+   int index = 0;
+   do {
+      int remainder = value % 16;
+      if (remainder < 10) hex[index++] = '0' + remainder;
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Use ternary operator.

File: Training24/NumConversion.c [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      input[strcspn (input, "\n")] = '\0';     // Remove newline character if present
+      if (IsValidInteger (input) && sscanf_s (input, "%d", number) == 1) return true;
+   }
+   return false;
+}
+
+/// <summary>Function to convert an integer into hexadecimal representation.</summary>
+void ConvertToHex (int number, char hex[], int bitWidth) {
+   int value;
+   int hexDigits = bitWidth / 4;   // Calculate number of hex digits based on bit width
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]:
```suggestion
int hexDigits = bitWidth / 4;
```
You should explain what this number 4 is using comments or use a variable/sizeof, so it is not difficult in the future to change this behavior if required

File: Training24/NumConversion.c [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+   if (fgets (input, sizeof (input), stdin) != NULL) {
+      input[strcspn (input, "\n")] = '\0';     // Remove newline character if present
+      if (IsValidInteger (input) && sscanf_s (input, "%d", number) == 1) return true;
+   }
+   return false;
+}
+
+/// <summary>Function to convert an integer into hexadecimal representation.</summary>
+void ConvertToHex (int number, char hex[], int bitWidth) {
+   int value;
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]:
```suggestion
int index = 0, hexDigits = bitWidth / 4, remainder = 0, 
    value = number < 0 ? (1 << bitWidth) + number : number;
```

File: Training24/NumConversion.c [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+   if (number < 0) value = (1 << bitWidth) + number;
+   else value = number;
+   int index = 0;
+   do {
+      int remainder = value % 16;
+      if (remainder < 10) hex[index++] = '0' + remainder;
+      else hex[index++] = 'A' + (remainder - 10);
+      value /= 16;
+   } while (value > 0);
+   while (index < hexDigits) hex[index++] = '0';   // Pad with leading zeros for the required bit width
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: You could also fill the array in reverse order, so you don't have to traverse it twice. int index = 0, hexDigits = bitWidth / 4, remainder = 0, 
    value = number < 0 ? (1 << bitWidth) + number : number;
```suggestion
hex[hexDigits] = '\0';
do {
   remainder = value % 16;
   hex[--hexDigits] = remainder < 10 ? '0' + remainder : 'A' + (remainder - 10);
   value /= 16;
} while (value > 0);
```

File: Training24/NumConversion.c [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+int DetermineBitWidth (int number) {   
+   if (number == INT_MIN) return 32;   // Determine the minimum bit width required to represent the number
+   if (number < -(1 << 15) || number >= (1 << 15)) return 32;
+   if (number < -(1 << 7) || number >= (1 << 7)) return 16;
+   return 8;   // Default bit width
+}
+
+int main () {
+   int number;
+   int binaryArray[32];
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: image
If I enter an extremely large number (that causes integer overflow, It considers it -1 and prints result. Better to handle that case. 

Also, you can declare multiple variables in the same statement. Please fix this everywhere in the program.

File: Training24/NumConversion.c [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+   while (true) {
+      printf ("Enter an integer: ");
+      if (GetValidInteger (&number)) break;
+      printf ("Invalid input. Please enter a valid integer.\n");
+   }
+
+   int bitWidth = DetermineBitWidth (number);
+   ConvertToBinary (number, binaryArray, bitWidth);
+   printf ("Binary: ");
+   PrintBinary (binaryArray, bitWidth);
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: You could make the program more intuitive by returning char* from the convert functions.
That way you could just write " printf ("Binary: %s ", ConvertToBinary (number, binaryArray, bitWidth)); "

File: Training24/NumConversion.c [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+#define BITS_PER_HEX_DIGIT 4
+#define MAX_INPUT_LENGTH 100
+#define MAX_HEX_DIGITS 32
+
+/// <summary> Function to convert an integer to binary and return it as a string</summary>
+char* ConvertToBinary (int number, int bitWidth) {
+   static char binaryString[33];
+   for (int i = 0; i < bitWidth; i++) {
+      binaryString[i] = ((number >> (bitWidth - i - 1)) & 1) ? '1' : '0';
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]:
```suggestion
      binaryString[i] = (number >> (bitWidth - i - 1)) & 1);

```

File: Training24/NumConversion.c [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+   }
+   binaryString[bitWidth] = '\0';
+   return binaryString;
+}
+
+
+/// <summary>Function to  check if input is a valid integer.</summary>
+bool IsValidInteger (const char* str) {
+   if (*str == '-' || *str == '+') str++;   // Skip sign if present
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Check for null string as you enter the function as well.

File: Training24/NumConversion.c [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+/// <summary>Function to determine the appropriate bit width based on the integer value.</summary>
+int DetermineBitWidth (int number) {   
+   if (number == INT_MIN) return 32;   // Determine the minimum bit width required to represent the number
+   if (number < -(1 << 15) || number >= (1 << 15)) return 32;
+   if (number < -(1 << 7) || number >= (1 << 7)) return 16;
+   return 8;   // Default bit width
+}
+
+int main () {
+   int number; char hex[MAX_HEX_DIGITS + 1];// Space for hex digits + null terminator
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Please write a test covering edge cases as well.

File: Training24/NumConversion.c [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+      if (!isdigit (*str)) return false; 
+      str++;
+   }
+   return true;
+}
+
+/// <summary>Function to get a valid integer from the user.</summary>
+bool GetValidInteger (int* number) {
+   char input[MAX_INPUT_LENGTH];
+   if (fgets (input, sizeof (input), stdin) != NULL) {
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]:
```suggestion
   if (fgets (input, sizeof (input), stdin) != NULL) return false;
```
See [guard](https://en.wikipedia.org/wiki/Guard_(computer_science)).

File: Training24/NumConversion.c [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+      input[strcspn (input, "\n")] = '\0'; // Remove newline character
+      if (IsValidInteger (input)) {
+         long long temp = 0;
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: avoid using long long unless necessary

File: Training24/NumConversion.c [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+   }
+   return true;
+}
+
+/// <summary>Function to get a valid integer from the user.</summary>
+bool GetValidInteger (int* number) {
+   char input[MAX_INPUT_LENGTH];
+   if (fgets (input, sizeof (input), stdin) != NULL) {
+      input[strcspn (input, "\n")] = '\0'; // Remove newline character
+      if (IsValidInteger (input)) {
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Similarly, 
```suggestion
      if (!IsValidInteger (input)) return false;
```

**Review by Dakshita-Minocha — COMMENTED** [review]
This is well written code. Good work!

File: Training24/NumConversion.c [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+   if (!*str) return false;   // If string is empty after sign
+   while (*str) {   
+      if (!isdigit (*str)) return false; 
+      str++;
+   }
+   return true;
+}
+
+/// <summary> Checks if the number is within the 32-bit integer range defined by INT_MIN and INT_MAX. </summary>
+bool IsWithinRange (long long number) {
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Here, you are checking for integer overflow. A better way to check would be to add 1 to the int and see if the result is less than the original int and vice versa for int_min. Also, the function name could be more indicative of what it actually does, for eg: IsWithinInt32Range / CheckOverflow

File: Training24/NumConversion.c [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+   input[strcspn (input, "\n")] = '\0';   // Remove newline character
+   if (strlen (input) == 0) return false;
+   char* num;    
+   *number = strtoll (input, &num, 10);
+   return (*num == '\0');
+}
+
+/// <summary>Function to convert an integer into hexadecimal representation.</summary>
+void BinaryToHex (const char binary[], char hex[], int bitWidth) {
+   int hexDigits = bitWidth / BITS_PER_HEX_DIGIT;
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]:
```suggestion
   int hexDigits = bitWidth / BITS_PER_HEX_DIGIT, binaryIndex = 0, n = 0;

```

File: Training24/NumConversion.c [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+   int binaryIndex = 0;
+   for (int i = 0; i < hexDigits; i++) {
+      int n = 0;
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c:61]:
```suggestion
      n = 0;

```

File: Training24/NumConversion.c [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+   if (number < -(1 << 15) || number >= (1 << 15)) return 32;
+   if (number < -(1 << 7) || number >= (1 << 7)) return 16;
+   return 8;   // Default bit width
+}
+
+/// <summary> Test binary and hexadecimal conversions with various inputs and checks for correctness.</summary>
+void TestCases () {
+   struct {
+      long long number;
+      char expectedBinary[MAX_BINARY_DIGITS]; char expectedHex[MAX_HEX_DIGITS + 1];
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Please note that if 2 or more variables are of the same type, you do not need to declare each of them in a separate statement. Fix this throughout your code.
```suggestion
      char expectedBinary[MAX_BINARY_DIGITS], expectedHex[MAX_HEX_DIGITS + 1];

```

File: Training24/NumConversion.c [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+   if (number == INT_MIN) return 32;   // Determine the minimum bit width required to represent the number
+   if (number < -(1 << 15) || number >= (1 << 15)) return 32;
+   if (number < -(1 << 7) || number >= (1 << 7)) return 16;
+   return 8;   // Default bit width
+}
+
+/// <summary> Test binary and hexadecimal conversions with various inputs and checks for correctness.</summary>
+void TestCases () {
+   struct {
+      long long number;
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c:79]: Check for ints.

File: Training24/NumConversion.c [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+      char expectedBinary[MAX_BINARY_DIGITS]; char expectedHex[MAX_HEX_DIGITS + 1];
+   } testCases[] = {
+       {0, "00000000", "00"},
+       {-128, "10000000", "80"},
+       {255, "0000000011111111", "00FF"},
+       {INT_MAX, "01111111111111111111111111111111", "7FFFFFFF"},
+       {INT_MIN, "10000000000000000000000000000000", "80000000"},
+       {2147483649ll, "00000000000000000000000000000000", "00000000"},
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: testcases is a struct for long long numbers. You are attempting to store long long numbers. This cast/conversion is not required.
See [here](https://stackoverflow.com/questions/16248221/what-is-1ll-or-2ll-in-c-and-c).

File: Training24/NumConversion.c [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+       {INT_MAX, "01111111111111111111111111111111", "7FFFFFFF"},
+       {INT_MIN, "10000000000000000000000000000000", "80000000"},
+       {2147483649ll, "00000000000000000000000000000000", "00000000"},   
+       {-2147483649ll, "10000000000000000000000000000000", "80000000"}   // Error cases: Out of range
+   };
+   int numTests = sizeof (testCases) / sizeof (testCases[0]);  
+   for (int i = 0; i < numTests; i++) {
+      long long number = testCases[i].number;
+      if (IsWithinRange (number)) {
+         int numberRange = (int)number;
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c:95]: If you check for overflow with integers by adding one, you will not need to do this.

File: Training24/NumConversion.c [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+         int bitWidth = DetermineBitWidth (numberRange);
+         char binary[MAX_BINARY_DIGITS]; char hex[MAX_HEX_DIGITS + 1];
+         ConvertToBinary (numberRange, bitWidth, binary);
+         BinaryToHex (binary, hex, bitWidth);
+         printf ("Number = %lld\n", number);
+         printf ("Binary value: %s\n", binary);
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]:
```suggestion
         printf ("Binary value: %s\nBinary Test: %s\n", binary, strcmp (binary, testCases[i].expectedBinary) == 0 ? GREEN"PASS" RESET : RED "FAIL" RESET);

```

Comment (Dakshita-Minocha) [reply Training24/NumConversion.c]: Similarly, you do not need to have a separate printf statement for every line. Improves readability.

File: Training24/NumConversion.c [hunk 30]
```diff
@@ -0,0 +1,141 @@
...
+         printf ("Binary Test: %s\n", strcmp (binary, testCases[i].expectedBinary) == 0 ? GREEN"PASS" RESET : RED "FAIL" RESET);
+         printf ("Hexadecimal value: %s\n", hex);
+         printf ("Hexadecimal Test: %s\n\n", strcmp (hex, testCases[i].expectedHex) == 0 ? GREEN "PASS" RESET : RED "FAIL" RESET);
+      }
+      else {
+         printf ("Number = %lld\n", number);
+         printf (YELLOW "Error: Number is out of 32-bit range.\n" RESET "\n");
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]:
```suggestion
printf ("Number = %lld\n"YELLOW "Error: Number is out of 32 - bit range.\n" RESET "\n", number);

```

File: Training24/NumConversion.c [hunk 31]
```diff
@@ -0,0 +1,141 @@
...
+   }
+}
+int main () {
+   long long number;
+   char binary[MAX_BINARY_DIGITS];
+   char hex[MAX_HEX_DIGITS + 1];
+   TestCases ();
+   while (true) {
+      printf ("Enter an integer: ");
+      if (GetValidInteger (&number)) {
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: I think it is enough to check for overflow within GetValidInteger(). If it is not withing Int32 range, it is invalid.

File: Training24/NumConversion.c [hunk 32]
```diff
@@ -0,0 +1,141 @@
...
+         if (IsWithinRange (number)) {
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]:
```suggestion
         if (!IsWithinRange (number)) printf (YELLOW "Error: Number is out of 32-bit range.\n" RESET "\n");

```
See [Guard](https://en.wikipedia.org/wiki/Guard_(computer_science)).

File: Training24/NumConversion.c [hunk 33]
```diff
@@ -0,0 +1,141 @@
...
+#define RED     "\033[1;31m"
+#define GREEN   "\033[1;32m"
+#define YELLOW  "\033[1;33m"
+#define MAGENTA "\033[1;35m"
+#define RESET   "\033[0m"
+
+/// <summary> Converts an integer to its binary representation as a string. </summary>
+void ConvertToBinary (int number, int bitWidth, char* binaryString) {
+   for (int i = 0; i < bitWidth; i++) binaryString[i] = ((number >> (bitWidth - i - 1)) & 1) ? '1' : '0';
+   binaryString[bitWidth] = '\0';  // Null-terminate the string
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Please include changes about comments from recent review.

File: Training24/NumConversion.c [hunk 34]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+/// <summary>Function to  check if input is a valid integer.</summary>
+bool IsValidInteger (const char* str) {
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Remove unnecessary code. This function is not being used anywhere.

File: Training24/NumConversion.c [hunk 35]
```diff
@@ -0,0 +1,141 @@
...
+   char input[MAX_INPUT_LENGTH];
+   if (fgets (input, sizeof (input), stdin) == NULL) return false;
+   input[strcspn (input, "\n")] = '\0';   // Remove newline character
+   if (strlen (input) == 0) return false;
+   char* num;    
+   *number = strtoll (input, &num, 10);
+   return (*num == '\0');
+}
+
+/// <summary>Function to convert an integer into hexadecimal representation.</summary>
```
Comment (Dakshita-Minocha) [inline Training24/NumConversion.c]: Your comment and function name both convey different uses for the function.

**Review by Dakshita-Minocha — APPROVED** [review]
